base/compiler/abstractinterpretation.jl

```diff
@@ -1925,16 +1916,21 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
                         ssavaluetypes[pc] = t
                     end
                 end
-                if frame.cur_hand !== nothing && isa(changes, StateUpdate)
-                    # propagate new type info to exception handler
-                    # the handling for Expr(:enter) propagates all changes from before the try/catch
-                    # so this only needs to propagate any changes
-                    l = frame.cur_hand.first::Int
-                    if stupdate1!(states[l]::VarTable, changes::StateUpdate) !== false
-                        if l < frame.pc´´
-                            frame.pc´´ = l
+                cur_hand = frame.handler_at[pc]
+                if isa(changes, StateUpdate)
+                    while cur_hand != 0
+                        let l = frame.handler_at[cur_hand + 1]
+                            # propagate new type info to exception handler
+                            # the handling for Expr(:enter) propagates all changes from before the try/catch
+                            # so this only needs to propagate any changes
+                            if stupdate1!(states[l]::VarTable, changes::StateUpdate) !== false
+                                if l < frame.pc´´
+                                    frame.pc´´ = l
+                                end
+                                push!(W, l)
+                            end
```

Comment:
Well, why do we want to propagate changes via `:enter` ? Can't we propagate changes directly to `:leave` statements here ?

Reply:
The `l` value here is short for `:leave` (strictly speaking, it doesn't have to be a :leave, but it normally will be)

Reply:
`frame.src.code[l]` is supposed to be `:enter` expression always, no ?
My understanding is that the algorithm accumulates all changes within `try/catch` clauses to `:enter` expression's states, and then `:enter` will propagate the changes to `:leave`.

So I wonder why we don't want to do:
```julia
if isa(changes, StateUpdate)
    while cur_hand != 0
        let l = frame.handler_at[cur_hand + 1]
            # propagate new type info to exception handler
            enter = frame.src.code[l]
            @assert isexpr(enter, :enter)
            leavestate = states[enter.args[1]::Int]::VarTable
            stupdate1!(leavestate, changes::StateUpdate) !== false
        end
        cur_hand = frame.handler_at[cur_hand]
    end
end
```

Reply:
Because `l` is almost always a `:leave` expr there

Reply:
oh, or was supposed to be, until I rearranged the code, and then it wasn't anymore